### ISSUE TITLE
Setting the primary scrolling direction for an scrollable element

### ISSUE THREAD

**Issue by Zhang-Junzhi** [w3c/csswg-drafts#10060]
Currently, almost all UAs assume the vertical direction is the first-class scrolling direction, so that UAs make vertical scroll as the _pure_ wheel event's default action, while making horizontal scroll as the wheel event's default action with an additional key(typically Shift Key).

That's a reasonable assumption for a page in horizontal writing page. A page in vertical writing mode, however, is another story, which is much more likely to overflow horizontally.

If we can set an element's primary scrolling direction, then the UAs know that it can make horizontal scroll as the _pure_ wheel event's default action.

**Comment by Zhang-Junzhi** [comment]
Also, there are two vertical writing modes: `vertical-rl` and `vertical-lr`.
It is worth suggesting UAs to "convert" scrolling direction **logically**.

**Comment by Zhang-Junzhi** [comment]
Just as a personal opinion:
I personally prefer using the concept of **overflow**, instead of **scroll**, for example, I would prefer a property named `primary-overflow-direction` over `primary-scroll-direction`.

Because `overflow` is more "style-oriented" name, although I started the topic using the concept of `scroll`, for the purpose of intuitively describing the issue.

**Comment by Zhang-Junzhi** [comment]
Just as a personal opinion:
I personally prefer using the concept of **overflow**, instead of **scroll**, for example, I would prefer a property named `primary-overflow-direction` over `primary-scroll-direction`.

Because **overflow** is more "style-oriented" name, although I started the topic using the concept of **scroll**, for the purpose of intuitively describing the issue.

**Comment by Zhang-Junzhi** [comment]
Just as a personal opinion:

I personally prefer the concept of **overflow**, instead of **scroll** for CSS, for example, I would prefer a property named `primary-overflow-direction` over `primary-scroll-direction`.

Because **overflow** is a more "style-oriented" name, although I started the topic using the concept of **scroll**, for the purpose of intuitively describing the issue.

**Comment by Zhang-Junzhi** [comment]
Further argument: Why is this feature necessary for CSS?

Although JavaScript developers are absolutely able to **programmatically** change the default action of a wheel event for vertical-writing pages, but they have to set the wheel event's `passive` option to `false`, in order to be able to call `preventDefault` to prevent the default action, and then write their own action code. Unfortunately, turning off `passive` mode results in a very big performance loss. Also, this couples style with event handling, making code hacky.

**Comment by fantasai** [comment]
I think UAs should be doing this automatically, based on the [principal writing mode](https://www.w3.org/TR/css-writing-modes-4/#principal-flow). So I guess there's a few questions:

- Do implementers agree with this interpretation?
- Should we clarify expectation this in the spec somehow, e.g. in an example or a note?
- Is there a way to write a WPT test for this?
- Are there bugs filed against implementations for this?

**Comment by Zhang-Junzhi** [comment]
> I think UAs should be doing this automatically, based on the [principal writing mode](https://www.w3.org/TR/css-writing-modes-4/#principal-flow).

It's absolutely reasonable to do this automatically based on the principal writing mode. However, as an alternate solution, I would also like to further introduce just my two cents on adding a new property (For the sake of concise description, I didn't mention the other reason behind adding a new property in my previous text):

Although what this issue mainly aims to handle is about vertical writing mode. But there might also be cases where it's desirable for pages even when it's principal writing mode is horizontal. For example, a `horizontal` page can use the property if it contains a bunch of extremely wide cityscape pictures:

```
<div style="primary-overflow-direction: horizontal; overflow: auto">
    <!--a bunch of extrememly wide cityscape pictures-->
    <picture ...>
    <picture ...>
    <picture ...>
    ...
</div>
```

>     * Are there bugs filed against implementations for this?

Yes, there're at least bugs filed for this kind of issues, as far as I know, on bugzilla.mozilla.org:

Years ago, [the issue was mentioned by David Baron](https://bugzilla.mozilla.org/show_bug.cgi?id=autodir), and then I implemented `autodir` feature for Firefox.

later on, [the Firefox Desktop team also utilise the `autodir` feature to improve UX](https://bugzilla.mozilla.orgshow_bug.cgi?id=1749825).

Unfortunately, the big disadvantage of `autodir` is that it's not a CSS thing, so it's unaware of any element's style. Turning `autodir` on by default would mean pages all over the world that contain any horizontally scrollable elements could be affected. So Firefox had to turn `autodir` off by default.

Things are much better if we solve this issue with CSS: the issue that `autodir` couldnot resolve can be resolved perfectly by CSS: Because with CSS, we can **selectively** affect pages or elements according to it's style(Whether automatically based on the principal writing mode, as you said, or manually set using a new property, as I said).

**Comment by fantasai** [comment]
> Turning autodir on by default would mean pages all over the world that contain any horizontally scrollable elements could be affected.

That seems... expected?

> So Firefox had to turn autodir off by default.

Can you explain why?

**Comment by SebastianZ** [comment]
> > Turning autodir on by default would mean pages all over the world that contain any horizontally scrollable elements could be affected.
> 
> That seems... expected?

Not necessarily. Imagine an article with a horizontal image gallery. If the reader is only interested in the article text, it would be surprising/annoying for them if while scrolling down the scroll direction suddenly changed within the image gallery.

Sebastian

**Comment by Zhang-Junzhi** [comment]
> > Turning autodir on by default would mean pages all over the world that contain any horizontally scrollable elements could be affected.
> 
> That seems... expected?
> 
> > So Firefox had to turn autodir off by default.
> 
> Can you explain why?

@SebastianZ gave a good example why `autodir` has to be turned off by default: Because `autodir` is unware of CSS style, so it will **blindly** apply to ANY elements if they are horizontally scrollable.
 
On the contrary, we can **selectively** affect pages or elements according to it's style(Either automatically based on the principal writing mode, as you said, or manually set using a new property, as I said). This is where a CSS solution will go beyond `autodir`.

**Comment by fahlen669** [comment]
This us fine meme ._.

**Comment by fantasai** [comment]
> Because autodir is unware of CSS style, so it will blindly apply to ANY elements if they are horizontally scrollable.

I don't understand. Wasn't the behavior taking into consideration both `overflow` and `writing-mode`?

> Not necessarily. Imagine an article with a horizontal image gallery. If the reader is only interested in the article text, it would be surprising/annoying for them if while scrolling down the scroll direction suddenly changed within the image gallery.

Suppose the gallery is vertical instead of horizontal. Why should that be different? Either the user wants to scroll through the image gallery or they don't. Laying out the gallery horizontally vs vertically shouldn't make a difference to whether scroll events are sent to the gallery or the main page.

**Comment by Zhang-Junzhi** [comment]
> I don't understand. Wasn't the behavior taking into consideration both `overflow` and `writing-mode`? These are both CSS properties.
> 

Sorry for being unclear. When I said "CSS unaware", I meant we hadn't introduced any CSS private property to do `autodir`, so `autodir` has nothing to do with CSS's extra assistance. Let me be more specific: `autodir` does "use" `writing-mode` and `overflow` indirectly(without the concept of CSS), in some way, in some condition. But that's not the key point. The key point is: **No selectiveness**. We cannot _selectively_ turn on `autodir` for some elements in the document, while turning off `autodir` for some other elements in the same document. Because we do not have such a CSS property that _selectively_ does it.

> Suppose the gallery is vertical instead of horizontal. Why should that be different? Either the user wants to scroll through the image gallery or they don't. Laying out the gallery horizontally vs vertically shouldn't make a difference to whether scroll events are sent to the gallery or the main page.

It does matter based on whether `autodir` is enabled.

```
<article class="very-tall"> <!--overflows vertically-->
    <img class="very-wide" ...> <!--overflows horizontally-->
</article>
```

If `autodir` is disabled, a user can scroll the `<article>` from top to down, using plain wheel scroll(plain means no Shift), without the possibility of accidentally falling into the <img>. Because plain wheel scroll will not be changed by disabled `autodir`, so, the `<img>` is "essentially an unscrollable element" from the perspective of plain wheel scroll.

If `autodir` is enabled, then there's possibility of falling into the `<img>`.

**Comment by Zhang-Junzhi** [comment]
Let me summarize my personal key solution for this topic in possibly a shortest sentence:

**All we need is selectiveness**.

```
<style>
.i-am-pretty-sure-users-will-be-interested
{
    primary-overflow-direction: horzizontal;
}
</style>

<article class="very-tall">
    <img class="very-wide i-am-not-sure-user-will-be-interested" ...>
</article>

<article class="very-tall">
    <img class="very-wide i-am-pretty-sure-users-will-be-interested" ...>
</article>
```

**Comment by Zhang-Junzhi** [comment]
BTW, just one more thing: A name just came to my mind: `more-interested-overflow-direction`, which I think reflects the intention of the property better than my previously proposed name `primary-overflow-direction`.

But `more-interested-overflow-direction` is awfully long due to my English proficiency. I believe someone can think of a better name.

**Comment by Zhang-Junzhi** [comment]
BTW, just one more thing: A name just came to my mind: `more-interesting-overflow-direction`, which I think reflects the intention of the property better than my previously proposed name `primary-overflow-direction`.

But `more-interesting-overflow-direction` is awfully long due to my English proficiency. I believe someone can think of a better name.

**Comment by GeeLaw** [comment]
Coming from #4380 :

Count me in and I think it's important to allow scroll axis to vary independently of the writing mode. My personal website has a multi-column layout for languages in LTR-TTB writing mode, and it should be scrolled horizontally.

There is another issue --- some mice allow four-direction scrolling. If rotating the wheel creates a horizontal scroll, what happens when the user side-clicks the wheel? I think only the wheel rotation interpretation is changed, so side-clicks also cause horizontal scroll. I checked with a Windows 8 app (Wolfram|Alpha) running on Windows 11, and my interpretation is what happens.

I like the concept of 1D and 2D scroll devices.

The distinction of 1D and 2D scroll device should be as follows: if the device gives equal access to two orthogonal directions of scroll, then it's a 2D scroll device, otherwise it's 1D.

- Mice whose wheel supports rotating and side-clicks: The rotate is much easier to access, provides finer scroll progression, than side-clicks (one click is a fixed, somewhat large scroll, and using clicks to scroll is much less smooth than wheels, even for very coarse-grained wheels). Therefore, this is a 1D scroll device.
- Precision touch pad, touch screen, non-precision touch pad with bottom & right edges for scrolling (see #4380 for reference photo of ThinkPad T400): These are 2D scroll devices.

---

Attempted proposal, combined with new names:

```css
* /* possible values */
{
  /* behavior interpreted by user agent */
  primary-overflow-axis: auto;

  /* axes dependent on writing-mode */
  primary-overflow-axis: block;
  primary-overflow-axis: inline;

  /* absolute axes */
  primary-overflow-axis: x;
  primary-overflow-axis: y;

  /* default on all elements except :root */
  primary-overflow-axis: inherit;
}

/* user agent stylesheet */
:root
{
  primary-overflow-axis: auto;
}
```

This CSS property should not change how `wheel` events are dispatched in response to wheel rotations --- it's always `deltaY` and positive means the user rotates the wheel towards oneself on Windows (in LTR-TTB and most browsers, this means the content below is made visible).

The reason to not change that is two-fold.

First, it'd be impossible to determine the raw readings if it were changed! Consider the case of wheel rotation and side-clicks. Both will convert to horizontal scroll if `primary-overflow-axis: x`, but JavaScript authors might want to do different things if they are handling `wheel`.

Second, there's a performance penalty to do that when firing `wheel` events. To determine the rotation axis, the user agent must find the scrolling container targeted by the event, **if the event is to cause a scroll**. However, if the event has a non-passive handler, there's a high chance the wheeling doesn't cause scroll by virtue (or curse) of `preventDefault`. The effort to determine the axis is wasted. (For passive handlers, we don't want the event data to be completely different from non-passive handlers for consistency.)

However, in order to assist JavaScript authors in understanding the consequence of `wheel`. We can propose to add additional (potentially lazily evaluated) properties that computes the action of the user agent if the event is not `prevent`ed`Default`.

---

Issue to be discussed:

- Suppose when the user rotates the wheel towards oneself, the browser usually brings content below into view. (This assumption is false if you turn on, deep breath, the so-called "natural scroll" on macOS --- which doesn't understand the difference between panning and wheel rotations.) In TTB-RTL (e.g., classical Chinese), what is the most intuitive result when the user rotates the wheel towards oneself?
  - Option 1: To scroll to the left (the content "later" is brought into view). This means we read "usual scroll to the bottom" as "scroll to the block-reading direction".
  - Option 2: To scroll to the right (the content "prior" is brough into view). This means we read "usual scroll to the bottom" as "scroll to the right".

To me, Option 2 is a slight win, but I did have a hard time deciding which is the more intuitive --- because I rarely read classical Chinese typeset TTB-RTL.

There are other [writing directions](https://www.omniglot.com/writing/direction.htm) including bottom-to-top. In an ideal world, research into the intuition of those writing system's users determines the most intuitive action. At the same time, I feel the dominance of Latin script in computing might have affected people's UI intuition.

If we want to do a research, a good starting point seems to be people in Taiwan and Japan. In Taiwan, textbooks are TTB-RTL and a book starts from the "back". So do mangas in Japan.

**Comment by alsoLut** [comment]
We need two new CSS properties:

1. axis-swap: Maps scroll/drag events from the X-axis to the Y-axis and vice versa. 

2. reverse-direction: reverses the scroll direction, causing upward scroll events to produce downward content movement and vice versa.

**Comment by simevidas** [comment]
I created an English web page that scrolls horizontally instead of vertically. You can see it at https://switch2.šime.eu. 

My case is unrelated to writing directions. I chose horizontal scrolling purely because of the type of content on the page. It’s a visualization that is full height and wider than the viewport.

Is this a valid use-case?

**Comment by Zhang-Junzhi** [comment]
> I created an English web page that scrolls horizontally instead of vertically. You can see it at [https://switch2.šime.eu](https://switch2.%C5%A1ime.eu).
> 
> My case is unrelated to writing directions. I chose horizontal scrolling purely because of the type of content on the page. It’s a visualization that is full height and wider than the viewport.
> 
> Is this a valid use-case?

Yes, this use case aligns perfectly with the original issue. We are generally talking about setting the primary scrolling direction, regardless of writing modes, although one major motivation is about writing modes.

If we only wanted to tie the primary scrolling direction to the writing mode by default, we wouldn’t need an additional CSS property. However, I believe we should also account for cases where the primary scrolling direction needs to be manually set independently of the writing mode. This is why I suggest adding a new CSS property.
